File: backend/app/schemas/booking.py
from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional, List
from enum import Enum

from app.schemas.event import EventResponse


class BookingStatusEnum(str, Enum):
    PENDING = "PENDING"
    CONFIRMED = "CONFIRMED"
    PAYMENT_FAILED = "PAYMENT_FAILED"
    CANCELLED = "CANCELLED"


class BookingCreate(BaseModel):
    event_id: int = Field(
        ...,
        description="Target event ID"
    )

    seat_ids: List[int] = Field(
        ...,
        min_length=1,
        description="List of selected seat IDs"
    )


class BookingStatusUpdate(BaseModel):
    status: BookingStatusEnum

class SeatSummary(BaseModel):
    id: int
    row: str
    number: int

    class Config:
        from_attributes = True


class VenueSummary(BaseModel):
    id: int
    name: str
    location: str

    class Config:
        from_attributes = True


class EventSummary(BaseModel):
    id: int
    title: str
    start_time: datetime

    venues: List[VenueSummary] = []

    class Config:
        from_attributes = True


class VenueResponse(BaseModel):
    id: int
    name: str
    location: str

    model_config = {
        "from_attributes": True
    }


class SeatResponse(BaseModel):
    id: int
    row: str
    number: int
    venue: VenueResponse

    model_config = {
        "from_attributes": True
    }

class UserSummary(BaseModel):
    username: str
    email: str
    phone_no: Optional[str] = None

    class Config:
        from_attributes = True

class BookingSeatResponse(BaseModel):
    id: int
    seat_id: int
    price: float
    seat: SeatResponse

    class Config:
        from_attributes = True

class BookingResponse(BaseModel):
    id: int
    user_id: int
    event_id: int
    total_price: float
    status: BookingStatusEnum
    created_at: datetime
    user: UserSummary | None = None
    event: EventResponse | None = None
    booking_seats: List[BookingSeatResponse]

    class Config:
        from_attributes = True

File: frontend/src/app/core/services/booking/booking.ts
import { HttpClient } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';

export interface BookingCreate {
  event_id: number;
  seat_ids: number[];
}

export interface VenueSummary {
  id: number;
  name: string;
  location: string;
}

export interface VenueResponse {
  id: number;
  name: string;
  location: string;
}

export interface BookingSeatResponse {
  id: number;
  seat_id: number;
  price: number;
  seat: {
    id: number;
    row: string;
    number: number;
    venue: VenueResponse;
  };
}

export interface UserSummary {
  username: string;
  email: string;
  phone_no: string;
}

export interface BookingResponse {
  id: number;
  user_id: number;
  user?: UserSummary;
  event_id: number;
  total_price: number;
  status: BookingStatus;
  created_at: string;
  event?: {
    id: number;
    title: string;
    description?: string;
    category: string;
    start_time: string;
    end_time: string;
    base_price: number | string;
    venues: VenueSummary[];
  };
  booking_seats: BookingSeatResponse[];
}

export type BookingStatus =
  | 'PENDING'
  | 'CONFIRMED'
  | 'CANCELLED';

export interface BookingStatusUpdate {
  status: BookingStatus;
}

@Injectable({
  providedIn: 'root',
})
export class BookingService {
  private readonly API_URL =
    'http://localhost:8000/api/v1/bookings';

  private readonly http = inject(HttpClient);

  createBooking(
    booking: BookingCreate
  ): Observable<BookingResponse> {
    return this.http.post<BookingResponse>(
      this.API_URL,
      booking
    );
  }

  getMyBookings(): Observable<BookingResponse[]> {
    return this.http.get<BookingResponse[]>(
      `${this.API_URL}/me`
    );
  }

  getAllBookings(
    status?: BookingStatus
  ): Observable<BookingResponse[]> {
    let url = `${this.API_URL}/admin`;

    if (status) {
      url += `?status=${status}`;
    }

    return this.http.get<BookingResponse[]>(url);
  }

  updateBookingStatus(
    bookingId: number,
    status: BookingStatus
  ): Observable<BookingResponse> {
    const bookingStatus: BookingStatusUpdate = {
      status,
    };

    return this.http.patch<BookingResponse>(
      `${this.API_URL}/admin/${bookingId}/status`,
      bookingStatus
    );
  }

  getBooking(
    bookingId: number
  ): Observable<BookingResponse> {
    return this.http.get<BookingResponse>(
      `${this.API_URL}/${bookingId}`
    );
  }

  cancelBooking(
    bookingId: number
  ): Observable<BookingResponse> {
    return this.http.patch<BookingResponse>(
      `${this.API_URL}/${bookingId}/cancel`,
      {}
    );
  }

  getBookedSeats(
    eventId: number,
    venueId: number
  ): Observable<number[]> {
    return this.http.get<number[]>(
      `${this.API_URL}/event/${eventId}/venue/${venueId}/booked-seats`
    );
  }
}